src/java/org/apache/cassandra/index/sai/disk/v1/SAICodecUtils.java

```diff
@@ -0,0 +1,265 @@
+
+import java.io.IOException;
+
+import org.apache.cassandra.index.sai.disk.format.Version;
+import org.apache.lucene.codecs.CodecUtil;
+import org.apache.lucene.index.CorruptIndexException;
+import org.apache.lucene.store.ChecksumIndexInput;
+import org.apache.lucene.store.DataInput;
+import org.apache.lucene.store.IndexInput;
+import org.apache.lucene.store.IndexOutput;
+
+import static org.apache.lucene.codecs.CodecUtil.CODEC_MAGIC;
+import static org.apache.lucene.codecs.CodecUtil.FOOTER_MAGIC;
+
+public class SAICodecUtils
+{
+    public static final String FOOTER_POINTER = "footerPointer";
+
+    public static void writeHeader(IndexOutput out) throws IOException
+    {
+        out.writeInt(CODEC_MAGIC);
+        out.writeString(Version.LATEST.toString());
+    }
+
+    public static void writeFooter(IndexOutput out) throws IOException
+    {
+        out.writeInt(FOOTER_MAGIC);
+        out.writeInt(0);
+        writeCRC(out);
+    }
+
+    public static void checkHeader(DataInput in) throws IOException
+    {
+        final int actualMagic = in.readInt();
+        if (actualMagic != CODEC_MAGIC)
+        {
+            throw new CorruptIndexException("codec header mismatch: actual header=" + actualMagic + " vs expected header=" + CODEC_MAGIC, in);
+        }
+        final Version actualVersion = Version.parse(in.readString());
+        if (!actualVersion.onOrAfter(Version.EARLIEST))
+        {
+            throw new IOException("Unsupported version: " + actualVersion);
+        }
+    }
+
+    public static void checkFooter(ChecksumIndexInput in) throws IOException
+    {
+        validateFooter(in, false);
+        long actualChecksum = in.getChecksum();
+        long expectedChecksum = readCRC(in);
+        if (expectedChecksum != actualChecksum)
+        {
+            throw new CorruptIndexException("checksum failed (hardware problem?) : expected=" + Long.toHexString(expectedChecksum) +
+                                            " actual=" + Long.toHexString(actualChecksum), in);
+        }
+    }
+
+    public static void validate(IndexInput input) throws IOException
+    {
+        checkHeader(input);
+        validateFooterAndResetPosition(input);
+    }
+
+    public static void validate(IndexInput input, long footerPointer) throws IOException
+    {
+        checkHeader(input);
+
+        long current = input.getFilePointer();
+        input.seek(footerPointer);
+        validateFooter(input, true);
+
+        input.seek(current);
+    }
+
+    public static void validateChecksum(IndexInput input) throws IOException
+    {
+        long position = input.getFilePointer();
+        long expected = CodecUtil.retrieveChecksum(input);
+
+        input.seek(position);
+        long actual = CodecUtil.checksumEntireFile(input);
+        if (expected != actual)
+            throw new CorruptIndexException("checksum failed (hardware problem?) : expected=" + Long.toHexString(expected) + " actual=" + Long.toHexString(actual), input);
```

Comment:
With the new test, I think this is the only case that doesn't have coverage...nice.

Reply:
Well, that should have been tested as well. I've put a test in for it and found that the code is wrong. We don't need to read the `expected` checksum here because the `CodecUtil.checksumEntireFile` does all this for us. The method simplifies to just calling `CodecUtil.checksumEntireFile`.